### PR TITLE
perf: Quick prefix validation check

### PR DESCRIPTION
**Pull request by ruben-arts**
This pr will add a hash of the lock file to the `EnvironmentFile` (`.pixi/envs/default/conda-meta/pixi`)

It will look like this:
```json
{
  "manifest_path": "/home/rarts/dev/pixi/pixi.toml",
  "environment_name": "default",
  "pixi_version": "0.34.0",
  "environment_lock_file_hash": "4f36ee620f10329d"
}
```
And that hash will be compared with the current lockfile on `pixi run` `pixi shell` `pixi shell-hook`.

### Profile result

```shell
❯ hyperfine "pixi run echo" "old-pixi run echo"
Benchmark 1: pixi run echo
  Time (mean ± σ):     381.6 ms ±  22.1 ms    [User: 193.8 ms, System: 246.3 ms]
  Range (min … max):   344.5 ms … 414.3 ms    10 runs
 
Benchmark 2: old-pixi run echo
  Time (mean ± σ):     868.2 ms ±  58.2 ms    [User: 480.0 ms, System: 557.0 ms]
  Range (min … max):   791.1 ms … 950.8 ms    10 runs
 
Summary
  pixi run echo ran
    2.28 ± 0.20 times faster than old-pixi run echo
```
> [!NOTE]  
> The remaining `381ms` is the activation which is fixed by https://github.com/prefix-dev/pixi/pull/2367

### UX
- It's turned on by default
- You can request a re-validate on `pixi run/shell/shell-hook` with `--revalidate`
- All commands designed to update the lock file or `pixi install` will always re-validate.

### TODO:

- [x] : Add tests: chosen python integration tests as I was to fed up with the extreme amount of time spent on writing tests in Rust
- [x] : use Enum instead of booleans. Using `UpdateMode::QuickValidate` and `UpdateMode::Revalidate`
- [x] : Document behavior
- [x] : Extend logic to cli. `--revalidate`

### CODE REVIEW

File: src/cli/install.rs [hunk 1]
```diff
@@ -52,7 +52,7 @@ pub async fn execute(args: Args) -> miette::Result<()> {
         let environment = project.environment_from_name_or_env_var(Some(env))?;
 
         // Update the prefix by installing all packages
-        update_prefix(&environment, args.lock_file_usage.into(), false).await?;
+        update_prefix(&environment, args.lock_file_usage.into(), false, true).await?;
```
Comment (ruben-arts) [inline src/cli/install.rs:55]: Change the booleans into an enum so it's readable

File: tests/integration/test_run_cli.py [hunk 2]
```diff
@@ -57,3 +57,108 @@ def test_run_in_shell(pixi: Path, tmp_path: Path) -> None:
...
+    )
+
+    # Validate creation of the pixi file with the hash
+    pixi_file = default_env_path(tmp_path).joinpath("conda-meta").joinpath("pixi")
+    assert pixi_file.exists()
+    assert "environment_lock_file_hash" in pixi_file.read_text()
+
+    # Break environment on purpose
+    dummy_a_meta_files = default_env_path(tmp_path).joinpath("conda-meta").joinpath("dummy-a*.json")
+    import glob
```
Comment (Hofer-Julian) [inline tests/integration/test_run_cli.py:88]:
```suggestion
```

Let's use the glob functionality provided by `pathlib` instead.
In general, I would also prefer top-level imports unless there's a good reason for it

**Review by Hofer-Julian — CHANGES_REQUESTED** [review]
Looks good! I can also reproduce the speed-up on my machine 😄

I added a few suggestions.

File: src/environment.rs [hunk 3]
```diff
@@ -142,42 +152,59 @@ fn create_prefix_location_file(environment_dir: &Path) -> miette::Result<()> {
...
+        if let Some(packages) = environment.packages(platform) {
+            for package in packages {
+                urls.push(package.url_or_path().into_owned().to_string())
+            }
+        }
+        urls.sort();
+
+        urls.hash(&mut hasher);
+
+        LockedEnvironmentHash(format!("{:x}", hasher.finish()))
```
[the remark this answers is not in the document]
Comment (ruben-arts) [reply src/environment.rs:183]: The reason is that that was what was used in the `EnvironmentHash`, but maybe it should be extended to also include "package hash" level detail. Ill give it a go as it was missed by me.

File: src/lock_file/update.rs [hunk 4]
```diff
@@ -127,22 +142,71 @@ impl<'p> LockFileDerivedData<'p> {
...
-        // Save an environment file to the environment directory
+    pub async fn prefix(
+        &mut self,
+        environment: &Environment<'p>,
+        update_mode: UpdateMode,
+    ) -> miette::Result<Prefix> {
+        // Check if the prefix is already up-to-date by validating the hash with the environment file
+        let hash = self.locked_environment_hash(environment)?;
+        if update_mode == UpdateMode::QuickValidate {
+            if let Ok(Some(environment_file)) = read_environment_file(&environment.dir()) {
```
[the remark this answers is not in the document]
Comment (ruben-arts) [reply src/lock_file/update.rs:168]: Smart!

File: tests/integration/test_run_cli.py [hunk 5]
```diff
@@ -57,3 +57,108 @@ def test_run_in_shell(pixi: Path, tmp_path: Path) -> None:
...
```
[the remark this answers is not in the document]
Comment (ruben-arts) [reply tests/integration/test_run_cli.py]: Fixed in [`1b2ade3` (#2400)](https://github.com/prefix-dev/pixi/pull/2400/commits/1b2ade3bf11712ed2cdd95ab78e1835c5b2067b1)

**Review by Hofer-Julian — APPROVED** [review]
Great work @ruben-arts!

**Pull request merged**